src/messages.c

```diff
@@ -0,0 +1,260 @@
+bool unmarshall_bind_packet(PgSocket *client, PktHdr *pkt, PgBindPacket *bind_packet)
+{
+	const uint8_t *ptr;
+	const char *portal;
+	const char *statement;
+
+	mbuf_rewind_reader(&pkt->data);
+
+	/* Skip first 5 bytes, because we skip the 'B' and the 4 bytes which are the length of the message */
+	if (!mbuf_get_bytes(&pkt->data, 5, &ptr))
+		goto failed;
+
+	if (!mbuf_get_string(&pkt->data, &portal))
+		goto failed;
+
+	if (!mbuf_get_string(&pkt->data, &statement))
+		goto failed;
+
+	bind_packet->len = pkt->len;
+	bind_packet->portal = portal;
+	bind_packet->name = statement;
+
+	return true;
+
+  failed:
+    disconnect_client(client, true, "broken Bind packet");
+	return false;
+}
+
+/* Unmarshall Describe packet into PgDescribePacket struct for further processing */
+bool unmarshall_describe_packet(PgSocket *client, PktHdr *pkt, PgDescribePacket *describe_packet)
+{
+	const uint8_t *ptr;
+	char describe;
+	const char *statement;
+
+	if (incomplete_pkt(pkt))
+		return false;
+
+	mbuf_rewind_reader(&pkt->data);
+
+	/* Skip first 5 bytes, because we skip the 'D' and the 4 bytes which are the length of the message */
+	if (!mbuf_get_bytes(&pkt->data, 5, &ptr))
+		goto failed;
+
+	if (!mbuf_get_char(&pkt->data, &describe))
+		goto failed;
+
+	if (!mbuf_get_string(&pkt->data, &statement))
+		goto failed;
+
+	describe_packet->type = describe;
+	describe_packet->name = statement;
+
+	return true;
+
+  failed:
+    disconnect_client(client, true, "broken Describe packet");
+	return false;
+}
+
+/* Unmarshall Close packet into PgClosePacket struct for further processing */
+bool unmarshall_close_packet(PgSocket *client, PktHdr *pkt, PgClosePacket *close_packet)
+{
+	const uint8_t *ptr;
+	char type;
+	const char *name;
+
+	if (incomplete_pkt(pkt))
+		return false;
+
+	mbuf_rewind_reader(&pkt->data);
+
+	if (!mbuf_get_bytes(&pkt->data, 5, &ptr))
+		goto failed;
+
+	if (!mbuf_get_char(&pkt->data, &type))
+		return true;
+
+	if (!mbuf_get_string(&pkt->data, &name))
+		name = "";
+
+	close_packet->type = type;
+	close_packet->name = name;
+
+	slog_noise(client, "unmarshall_close_packet: type=%c, len=%d, S/P=%c, name=%s", pkt->type, pkt->len, type, name);
+
+	return true;
+
+  failed:
+    disconnect_client(client, true, "broken Close packet");
+	return false;
+}
+
+bool is_close_statement_packet(PgClosePacket *close_packet)
+{
+	return close_packet->type == 'S' && *close_packet->name;
+}
+
+PktBuf *create_parse_packet(char *statement, PgParsePacket *pkt)
+{
+	PktBuf *buf;
+	buf = pktbuf_dynamic(5 + strlen(pkt->query) + strlen(statement) + 2 + pkt->num_parameters * 4);
+	pktbuf_start_packet(buf, 'P');
+	pktbuf_put_string(buf, statement);
+	pktbuf_put_string(buf, pkt->query);
+	pktbuf_put_uint16(buf, pkt->num_parameters);
+	pktbuf_put_bytes(buf, pkt->parameter_types_bytes, pkt->num_parameters * 4);
+	pktbuf_finish_packet(buf);
+	return buf;
+}
+
+PktBuf *create_parse_complete_packet(void)
+{
+	PktBuf *buf;
+	buf = pktbuf_dynamic(5);
```

Comment:
Most of these can use pktbuf_static instead.

Reply:
Created buffer is returned by the function - no way to use  `pktbuf_static` here.
Certainly it is possible to rewrite code, so that buffer is sent in the same function.
But it is non-trivial refactoring.
Not sure if it actually needed.

Reply:
Looking a bit more at other code that sends packets. Those places (almost) always use the `pktbuf_write_generic`, `pktbuf_write_SomePacketType` and/or `SEND_SomePacketType` macros. I think even for consistency it's worth to refactor to replace usage of these functions with similar macros.

Reply:
Done

Reply:
Thanks, I wanted to take the refactoring a bit further. But I'll change that myself in a few minutes.